File: src/main/java/macro303/neptunes/technology/TechnologyAdapter.java
package macro303.neptunes.technology;

import com.google.gson.*;

import java.lang.reflect.Type;

/**
 * Created by Macro303 on 2018-05-10.
 */
public class TechnologyAdapter implements JsonSerializer<Technology>, JsonDeserializer<Technology> {

	@SuppressWarnings("unchecked")
	@Override
	public Technology deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
		var jsonObject = json.getAsJsonObject();

		var level = jsonObject.get("level").getAsInt();
		var value = jsonObject.get("value").getAsInt();

		return new Technology(level, value);
	}

	@Override
	public JsonElement serialize(Technology src, Type typeOfSrc, JsonSerializationContext context) {
		var jsonObject = new JsonObject();

		jsonObject.addProperty("level", src.getLevel());
		jsonObject.addProperty("value", src.getValue());

		return jsonObject;
	}
}

File: src/main/kotlin/macro/neptunes/technology/TechnologyAdapter.kt
package macro.neptunes.technology

import com.google.gson.*

import java.lang.reflect.Type

/**
 * Created by Macro303 on 2018-05-10.
 */
class TechnologyAdapter : JsonSerializer<Technology>, JsonDeserializer<Technology> {

	@Throws(JsonParseException::class)
	override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Technology {
		val jsonObject = json.asJsonObject

		val level = jsonObject.get("level").asInt
		val value = jsonObject.get("value").asDouble

		return Technology(level, value)
	}

	override fun serialize(src: Technology, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
		val jsonObject = JsonObject()

		jsonObject.addProperty("level", src.level)
		jsonObject.addProperty("value", src.value)

		return jsonObject
	}
}